src/main/java/org/jvnet/hudson/test/MemoryAssert.java

```diff
@@ -138,20 +145,74 @@ public static void assertHeapUsage(Object o, int max) throws Exception {
+        System.err.println("Trying to collect " + reference.get() + "…");
         Set<Object[]> objects = new HashSet<Object[]>();
-        while (true) {
+        int size = 1024;
+        while (reference.get() != null) {
             try {
-                objects.add(new Object[1024]);
+                objects.add(new Object[size]);
+                size *= 1.3;
             } catch (OutOfMemoryError ignore) {
                 break;
             }
+            System.gc();
+            System.err.println("GC after allocation of size " + size);
         }
         objects = null;
         System.gc();
         Object obj = reference.get();
-        if (obj != null) {
-            fail(LiveReferences.fromRoots(Collections.singleton(obj)).toString());
+        if (obj == null) {
+            System.err.println("Successfully collected.");
+        } else {
+            System.err.println("Failed to collect " + obj + ", looking for strong references…");
+            Map<Object,Path> rootRefs = LiveReferences.fromRoots(Collections.singleton(obj));
+            if (!rootRefs.isEmpty()) {
+                fail(rootRefs.toString());
+            } else {
+                System.err.println("Did not find any strong references to " + obj + ", looking for soft references…");
+                rootRefs = fromRoots(Collections.singleton(obj), null, null, new Filter() {
+                    final Field referent;
+                    {
+                        try {
+                            referent = Reference.class.getDeclaredField("referent");
+                        } catch (NoSuchFieldException x) {
+                            throw new AssertionError(x);
+                        }
+                    }
+                    @Override public boolean accept(Object obj, Object referredFrom, Field reference) {
+                        return !referent.equals(reference) || !(referredFrom instanceof WeakReference);
+                    }
+                });
+                if (!rootRefs.isEmpty()) {
+                    fail(rootRefs.toString());
+                } else {
+                    System.err.println("Did not find any soft references to " + obj + ", looking for weak references…");
+                    rootRefs = fromRoots(Collections.singleton(obj), null, null, ScannerUtils.skipObjectsFilter(Collections.<Object>singleton(reference), true));
+                    if (!rootRefs.isEmpty()) {
+                        fail(rootRefs.toString());
+                    } else {
+                        fail("Did not find any root references to " + obj + " whatsoever. Unclear why it is not being collected.");
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * TODO {@link LiveReferences#fromRoots(Collection, Set, BoundedRangeModel, Filter) logically ANDs the {@link Filter}
+     * with {@link ScannerUtils#skipNonStrongReferencesFilter}, making it useless for our purposes.
```

Comment:
Still no usable API for this in [current NB sources](http://hg.netbeans.org/main-silver/file/222964529b79/o.n.insane/src/org/netbeans/insane/impl/LiveEngine.java#l82).